### PR TITLE
[bitnami/elasticsearch] Add Curator

### PR DESCRIPTION
**Pull request by marcosbc**
**Description of the change**

Adds Curator support to Elasticsearch chart.

**Benefits**

It is now possible to use Curator to perform day-to-day maintenance and management on Elasticsearch clusters.

**Possible drawbacks**

None

None

**Applicable issues**

None

**Additional information**

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/elasticsearch/templates/configmap-es.yaml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{{- if .Values.config }}
+{{- if or .Values.config }}
```
Comment (carrodher) [inline bitnami/elasticsearch/templates/configmap-es.yaml]: `or`??

**Comment by carrodher** [comment]
I bumped the version, the merge integration checks failed

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  labels:
+    app: {{ template "elasticsearch.name" . }}
+    chart: {{ template "elasticsearch.chart" . }}
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    role: "curator"
+  name: {{ template "elasticsearch.curator.fullname" . }}-psp
+spec:
+  privileged: true
+  #requiredDropCapabilities:
```
Comment (tompizmor) [inline bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: Is this line commented intentionally?

File: bitnami/elasticsearch/README.md [hunk 3]
```diff
@@ -171,6 +171,37 @@ The following table lists the configurable parameters of the Elasticsearch chart
...
+| `curator.cronjob.jobRestartPolicy`                | Control the Job restartPolicy                                                                                                                             | `Never`                                                 |
+| `curator.podAnnotations`                          | Annotations to add to the pod                                                                                                                             | {}                                                      |
+| `curator.rbac.enabled`                            | Enable RBAC resources                                                                                                                                     | `false`                                                 |
+| `curator.serviceAccount.create`                   | Create a default serviceaccount for elasticsearch curator                                                                                                 | `true`                                                  |
```
Comment (juan131) [inline bitnami/elasticsearch/README.md]: Shouldn't we use `curator.rbac.enabled` to decide whether creating the serviceAccount or not the serviceAccount when is created?

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  volumes:
+    - 'configMap'
+    - 'secret'
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: This is only being applied to Curator due to the `heritage` label (currently set to `curator`)

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  supplementalGroups:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
+  fsGroup:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: Actually I just set `runAsUser` to a `MustRunAsNonRoot`. The other properties should be OK as is (i.e. check NGINX Ingress Controller).

File: bitnami/elasticsearch/README.md [hunk 6]
```diff
@@ -171,6 +171,37 @@ The following table lists the configurable parameters of the Elasticsearch chart
...
 | `ingest.readinessProbe.successThreshold`          | Minimum consecutive successes for the probe to be considered successful after having failed (ingest nodes pod)                                            | `1`                                                     |
 | `ingest.readinessProbe.failureThreshold`          | Minimum consecutive failures for the probe to be considered failed after having succeeded                                                                 | `5`                                                     |
+| `curator.enabled`                                 | Enable Elasticsearch Curator cron job                                                                                                                     | `false`                                                 |
+| `curator.name`                                    | Elasticsearch Curator pod name                                                                                                                            | `curator`                                               |
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/README.md]: Good catch, it should be used.

File: bitnami/elasticsearch/templates/NOTES.txt [hunk 7]
```diff
@@ -59,6 +59,17 @@
 
 ** Please be patient while the chart is being deployed **
 
+{{- if .Values.curator.enabled }}
+
+  A CronJob will run with schedule {{ .Values.curator.cronjob.schedule }}.
+
+  The Jobs will not be removed automagically when deleting this Helm chart.
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/NOTES.txt]: This is based from the official chart, I don't dislike it. :-)

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml (hunk 2 above)
Comment (marcosbc) [reply bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: Yes

File: bitnami/elasticsearch/templates/_helpers.tpl [hunk 8]
```diff
@@ -168,22 +168,25 @@ imagePullSecrets:
 {{- range .Values.global.imagePullSecrets }}
   - name: {{ . }}
 {{- end }}
-{{- else if or .Values.image.pullSecrets .Values.metrics.image.pullSecrets .Values.sysctlImage.pullSecrets .Values.volumePermissions.image.pullSecrets }}
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/_helpers.tpl]: OK, I've changed it. Thanks!

File: bitnami/elasticsearch/README.md (hunk 3 above)
Comment (marcosbc) [reply bitnami/elasticsearch/README.md]: I don't think so. You might want to use an existing service account. Also, this is similar to what we do in other charts (i.e. Metrics Server, NGINX Ingress Controller, Fluentd, MariaDB Galera...). I wouldn't change it.

File: bitnami/elasticsearch/templates/rolebinding.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  labels:
+    app: {{ template "elasticsearch.name" . }}
+    chart: {{ template "elasticsearch.chart" . }}
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    role: "curator"
+    component: elasticsearch-curator-configmap
+  name: {{ template "elasticsearch.curator.name" . }}-rolebinding
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/rolebinding.yaml]: done

File: bitnami/elasticsearch/templates/configmap-curator.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.curator.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "elasticsearch.curator.fullname" . }}-config
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/configmap-curator.yaml]: done

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: 👍

File: bitnami/elasticsearch/templates/podsecuritypolicy.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/elasticsearch/templates/podsecuritypolicy.yaml]: 👍

File: bitnami/elasticsearch/templates/NOTES.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/elasticsearch/templates/NOTES.txt]: I don't consider it serious

File: bitnami/elasticsearch/templates/cronjob.yaml [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  failedJobsHistoryLimit: {{ . }}
+  {{- end }}
+  {{- with .Values.curator.cronjob.successfulJobsHistoryLimit }}
+  successfulJobsHistoryLimit: {{ . }}
+  {{- end }}
+  jobTemplate:
+    metadata:
+      labels:
+        app: {{ template "elasticsearch.name" . }}
+        release: {{ .Release.Name }}
```
Comment (juan131) [inline bitnami/elasticsearch/templates/cronjob.yaml]: At `metadata.labels` you're using:
```
    release: {{ .Release.Name | quote }}
```
Here you're using:
```
        release: {{ .Release.Name }}
```
I don't mind if you quote the labels or not, but be consistent.

File: bitnami/elasticsearch/templates/cronjob.yaml [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+              volumeMounts:
+                - name: config-volume
+                  mountPath: /etc/es-curator
+{{- if .Values.curator.extraVolumeMounts }}
+{{ toYaml .Values.curator.extraVolumeMounts | indent 16 }}
+{{ end }}
+{{ if .Values.curator.command }}
+              command:
+{{ toYaml .Values.curator.command | indent 16 }}
+{{- end }}
```
Comment (juan131) [inline bitnami/elasticsearch/templates/cronjob.yaml]: Suggestion:
```diff
 {{ if .Values.curator.command }}
-              command:
-{{ toYaml .Values.curator.command | indent 16 }}
+              command: {{ toYaml .Values.curator.command | nindent 16 }}
 {{- end }}
```

File: bitnami/elasticsearch/templates/cronjob.yaml [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+{{- range $key,$value := .Values.curator.envFromSecrets }}
+              - name: {{ $key | upper | quote}}
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ $value.from.secret | quote}}
+                    key: {{ $value.from.key | quote}}
+{{- end }}
+{{- end }}
+              resources:
+{{ toYaml .Values.curator.resources | indent 16 }}
```
Comment (juan131) [inline bitnami/elasticsearch/templates/cronjob.yaml]: Suggestion:
```diff
-              resources:
-{{ toYaml .Values.curator.resources | indent 16 }}
+{{- if .Values.curator.resources }}
+              resources: {{ toYaml .Values.curator.resources | nindent 16 }}
+{{- end }}
```

File: bitnami/elasticsearch/templates/_helpers.tpl [hunk 14]
```diff
@@ -257,3 +232,83 @@ but Helm 2.9 and 2.10 does not support it, so we need to implement this if-else
...
+{{- end -}}
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "elasticsearch.curator.fullname" -}}
+{{- $name := default .Chart.Name .Values.nameOverride -}}
+{{- printf "%s-%s-%s" .Release.Name $name .Values.curator.name | trunc 63 | trimSuffix "-" -}}
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/_helpers.tpl]: We do it for others (i.e. Metrics). It also allows users to customize the name.

File: bitnami/elasticsearch/templates/cronjob.yaml [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    spec:
+      template:
+        metadata:
+          labels:
+            app: {{ template "elasticsearch.name" . }}
+            release: {{ .Release.Name }}
+{{- if .Values.podAnnotations }}
+          annotations:
+{{ toYaml .Values.podAnnotations | indent 12 }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (marcosbc) [reply bitnami/elasticsearch/templates/cronjob.yaml]: Good catch, I checked again for similar issues and fixed a few.

File: bitnami/elasticsearch/templates/cronjob.yaml (hunk 13 above)
Comment (marcosbc) [reply bitnami/elasticsearch/templates/cronjob.yaml]: OK! Changed.

File: bitnami/elasticsearch/templates/_helpers.tpl (hunk 14 above)
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/elasticsearch/templates/_helpers.tpl]: 🤷‍♂ I don't find it useful, but... I don't think it's an issue either

File: bitnami/elasticsearch/templates/cronjob.yaml (hunk 12 above)
Comment (juan131) [reply bitnami/elasticsearch/templates/cronjob.yaml]: Simplicity

**Pull request merged**